### PR TITLE
.travis: Run tests for py up to 3.4 only on master

### PR DESCRIPTION
**Pull request by Asalle**
Closes https://github.com/coala/coala/issues/4323

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -66,10 +66,20 @@ before_script:
   - python .misc/check_setuptools.py
 
 script:
-  - py.test
-  - python setup.py bdist_wheel
-  - pip install ./dist/coala-*.whl
-  - pip install coala-bears[alldeps] --pre -U
+  - >
+    if [ ${UNSUPPORTED} == "true"]; then
```
Comment (yukiisbored) [inline .travis.yml]: I think this should be it's own script. With https://github.com/coala/coala/pull/4704/files#diff-354f30a63fb0907d4ad57269548329e3R72 being a function.

**Comment by Asalle** [comment]
@yukiisbored what do you think about my changes?

File: .misc/build_and_test.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+set -e
+
+# Get environment variables, readily decrypted by rultor
+#source ../rultor_secrets.sh
+
+if [${UNSUPPORTED} == "false"] || [${TRAVIS_BRANCH} == "master"]; then
```
Comment (Makman2) [inline .misc/build_and_test.sh]: Below the if-conditions effectively form an AND/`&&`, now you use OR/`||`. Is this the same?

File: .misc/build_and_test.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (Makman2) [inline .misc/build_and_test.sh]: Because this file is bound to travis, the filename should maybe contain "travis" somehow

**Comment by Makman2** [comment]
Commit message:
> .travis: Run tests up to 3.4 only on master ~~only~~

;)

**Comment by Makman2** [comment]
Recommendation: I would switch the commit order: First move the old testing steps into a separate script, and then make the relevant modifications for the actual issue. This is easier I believe :+1: But optional 👍

**Comment by jayvdb** [comment]
Yup, agree with @Makman2's review comments above, including copy (without as little modification as possible) as one commit, then edit as separate commit.  That makes it easier to understand what happened.

File: .misc/build_and_test.sh (hunk 2 above)
Comment (RaiVaibhav) [inline .misc/build_and_test.sh]: I think this has to be and/`&&` condition given on issue if `[${TRAVIS_BRANCH} != "master"] then also it will run

Comment (RaiVaibhav) [reply .misc/build_and_test.sh]: but we want only on master

Comment (Asalle) [reply .misc/build_and_test.sh]: Yup, it is the same

Comment (Makman2) [reply .misc/build_and_test.sh]: can you explain me why? :) Because I still believe this isn't the same :)

Comment (Asalle) [reply .misc/build_and_test.sh]: Here's why I think it's equivalent:

1) In the previous commit we had this:

	   if [ ${UNSUPPORTED} == "true"]; then
	     if [ ${TRAVIS_BRANCH} == "master" ]; then
	       py.test;
	       python setup.py bdist_wheel;
	       pip install ./dist/coala-*.whl;
	       pip install coala-bears[alldeps] --pre -U;
	     fi
	   else
	     py.test;
	     python setup.py bdist_wheel;
	     pip install ./dist/coala-*.whl;
	     pip install coala-bears[alldeps] --pre -U;
	   fi

there are two conditions, 2^n cases could happen

	unsupported == true | 0 0 1 1
	branch == master    | 0 1 0 1

As you can see 
in (0, 0) (python >= 3.4, not master) we go staight to `else`, i. e. setup & run tests,
in (0, 1) (python >= 3.4, master) we again go to else, setup & run tests,
in (1, 0) (python < 3.4, not master) don't do anything and 
in (1, 1) (python < 3.4, master) we go to first build-and-test block

I think it needs refactoring, but still corresponds with somethings you would want to see.

2) I the last commint the previous code was changed to:

		if [${UNSUPPORTED} == "false"] || [${TRAVIS_BRANCH} == "master"]; then
		    cd ..
		    py.test;
		    python setup.py bdist_wheel;
		    pip install ./dist/coala-*.whl;
		    pip install coala-bears[alldeps] --pre -U;

		    # https://github.com/coala/coala-bears/issues/1037
		    if [[ "$TRAVIS_PULL_REQUEST" != "false" ]]; then
		      sed -i.bak '/bears = GitCommitBear/d' .coafile
		    fi

		    coala --non-interactive
		    python setup.py docs
		fi

let's create the same table for this predicate (btw, can I call it a predicate? xD)

	unsupported == true | 0 0 1 1
	branch == master    | 0 1 0 1

This is slightly annoying, because we changed unsupported == true to the opposite, 
but still:
(0, 0) (not unsupported, not master) -> build & test
(0, 1) (not unsupported, master)     -> build & test
(1, 0) (unsupported, not master) 	 -> do nothing
(1, 1) (unsupported, master) 		 -> build & test


So, in my point of view those two pieces code are equivalent, the only thing, 
that differs is that I moved running coala and generating docs inside the if,
because I thought, that if we don't need to run tests, we don't need docs.
I can reverse it, if you want it back.

Comment (Asalle) [reply .misc/build_and_test.sh]: We don't build only if we're not on master and we are not supported, 
let x == 'is not supported'
y == 'is on master'

`x && !y == don't build`

let's see when we do build

`!(x && !y) --by deMorgan theorem--> !x || y`

Comment (Asalle) [reply .misc/build_and_test.sh]: Also, look at my reply to RaiVaibhav :)

Can we drop building and buidling docs on python < 3.4?

File: .misc/build_and_test.sh (hunk 3 above)
Comment (Asalle) [reply .misc/build_and_test.sh]: Accepted

File: .misc/build_and_test.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# this is helper script for travis
+# it checks if we should run tests
```
Comment (Makman2) [inline .misc/build_and_test.sh]: not really, it actually executes also all test steps in the right cases. If you want to mention this, do it directly above the relevant if-clause :)

File: .misc/build_and_test.sh (hunk 2 above)
Comment (Makman2) [reply .misc/build_and_test.sh]: Indeed, good old boolean functions.

> Can we drop building and buidling docs on python < 3.4?

File an issue ;)

**Comment by Makman2** [comment]
Please make two commits out of this: one putting the existing build steps into a new file, and another one modifying it to conform to the issue and implementing the skip of builds on non-masters.

**Comment by Makman2** [comment]
unack 87e77e1

**Comment by Asalle** [comment]
@Makman2 hey, is it ok? :)

**Comment by Makman2** [comment]
> build_and_test: Rename to be more travis 

This one should be squashed with the commit creating the file, no need to do a rename when you already introduce the file here ;)

File: .misc/travis_build_and_test.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+set -e
+
+if [${UNSUPPORTED} == "false"] || [${TRAVIS_BRANCH} == "master"]; then
```
Comment (Mixih) [inline .misc/travis_build_and_test.sh]: I'm pretty sure you need spaces after and before the `]` and `[`

Comment (Mixih) [reply .misc/travis_build_and_test.sh]: see https://travis-ci.org/coala/coala/jobs/301516951#L688

Comment (Mixih) [reply .misc/travis_build_and_test.sh]: also consider using `[[`. it's generally safer for bash

Comment (Asalle) [reply .misc/travis_build_and_test.sh]: Hey, @Mixih why does by build fail when I change `[` to `[[`?

**Comment by Asalle** [comment]
@Mixih okay now? :)

@Makman2 appveyor is failing for reasons that are ouside the scope of this issue, what should I do, wait? (see https://github.com/coala/coala/pull/4869)

**Comment by RaiVaibhav** [comment]
https://github.com/coala/coala/issues/4870

**Comment by Makman2** [comment]
Whee that looks bad... Yeah let's try to wait, though my first intention tells me that `freezegun` might have been broken and we maybe need to limit its version...

File: .misc/travis_build_and_test.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+py.test
+python setup.py bdist_wheel
+pip install ./dist/coala-*.whl
+pip install coala-bears[alldeps] --pre -U
+    # https://github.com/coala/coala-bears/issues/1037
+    if [[ "$TRAVIS_PULL_REQUEST" != "false" ]]; then
+      sed -i.bak '/bears = GitCommitBear/d' .coafile
+    fi
```
Comment (Makman2) [inline .misc/travis_build_and_test.sh]: why the indentation? I think only the interior of the `if` should be indented^^

**Comment by Makman2** [comment]
About `.travis.yml: Move script section to separate file`

You maybe want to add a commit body about the motivation of this change. But optional, you don't have to.

File: .misc/travis_build_and_test.sh [hunk 7]
```diff
@@ -1,12 +1,19 @@
...
-    fi
+set -e
+
+if [[ ${UNSUPPORTED} == "false" || ${TRAVIS_BRANCH} == "master" ]]; then
+    py.test;
+fi
+
+python setup.py bdist_wheel;
+pip install ./dist/coala-*.whl;
+pip install coala-bears[alldeps] --pre -U;
```
Comment (Makman2) [inline .misc/travis_build_and_test.sh]: unnecessary semicolons at end, please remove all ;)

**Comment by Asalle** [comment]
@Makman2 okay now? :)

**Comment by yukiisbored** [comment]
LGTM.

ack https://github.com/coala/coala/pull/4704/commits/b9752078b836a53ac8c8ced97275464d6df756ea https://github.com/coala/coala/pull/4704/commits/6f65ebd8f6b006455803f799f3d03e99b0f469cd

**Comment by yukiisbored** [comment]
Gonna wait for @Makman2, just in case he wants to review it again.

**Comment by Asalle** [comment]
Hey @yukiisbored @Makman2 @RaiVaibhav can somebody merge this already? xD

File: .misc/travis_build_and_test.sh (hunk 6 above)
Comment (Makman2) [reply .misc/travis_build_and_test.sh]: bump

**Comment by Makman2** [comment]
Not all comments addressed, see https://github.com/coala/coala/pull/4704/commits/6f65ebd8f6b006455803f799f3d03e99b0f469cd#r151215451 ;)

**Comment by Makman2** [comment]
unack 6f65ebd

**Comment by Makman2** [comment]
👉 https://github.com/coala/coala/pull/4704/commits/24c2404fb85b67705d96723bd7a626000c6ab44c#r153625382

**Comment by Makman2** [comment]
unack 24c2404

**Comment by Asalle** [comment]
Hey @Makman2 I don't quite understand what do you mean. I re-read all your comments, and I thought that you mean, that I should add some commit description, but it's not it. What is it, then?

**Comment by Makman2** [comment]
Oh damn, seems GitHub doesn't link to it properly. However, you have indented the `if` block itself in the first commit, which should be aligned properly^^

**Comment by Makman2** [comment]
Maybe you can see the comment when you click on your commit again :+1:

**Comment by Asalle** [comment]
@Makman2 okay now?

File: .misc/travis_build_and_test.sh [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+python setup.py bdist_wheel
+pip install ./dist/coala-*.whl
+pip install coala-bears[alldeps] --pre -U
+
+# https://github.com/coala/coala-bears/issues/1037
+if [[ "$TRAVIS_PULL_REQUEST" != "false" ]]; then
+  sed -i.bak '/bears = GitCommitBear/d' .coafile
+fi
+
+>>>>>>> 412e04d9... temp
```
Comment (Makman2) [inline .misc/travis_build_and_test.sh]: git conflict marker ;)

File: .misc/travis_build_and_test.sh [hunk 9]
```diff
@@ -1,6 +1,11 @@
 #!/bin/bash
 
-py.test
+set -e
```
Comment (Makman2) [inline .misc/travis_build_and_test.sh]: maybe `set -e` should go into a separate commit? Or maybe into the previous one? Up to you :+1:

**Comment by Makman2** [comment]
Commit message of `travis_build_and_test: Test on master or py >= 3.4`
> We don't need tests to be run for python <= 3.3,
unless we are on master.
 ~~So we move the whole "script" section of travis.yml
to a separate file (travis_build_and_test.sh) and
run tests only if python > 3.3 or if we are on master.~~

This part can be removed :)

**Comment by Asalle** [comment]
@Makman2 updated

**Comment by Makman2** [comment]
> We need to move the original script section of the .travis.yml to a
separate file to stay clear, where the travis_build_and_test.sh
came from.

I don't like the reasoning. The reason is that it's cleaner and easier to maintain if we move the travis CI instructions into a separate file. Please incorporate that into the commit body :+1:

**Comment by Makman2** [comment]
ack 44240f3

**Comment by Makman2** [comment]
unack 08f1f18

**Comment by jayvdb** [comment]
This doesnt fix https://github.com/coala/coala/issues/4323 .  It doesnt even come close.

**Comment by Asalle** [comment]
@jayvdb in such case, please tell me, what do I understand wrong?

**Comment by Asalle** [comment]
@Makman2 updated.

**Comment by Makman2** [comment]
I would like to hear @jayvdb 's solution idea :)

**Comment by Mixih** [comment]
Hmm I think we misinterpreted the issue here. We do want to test on non-master branches (which this pr currently removes), but we don't want to run the python 2 and 3.3 tests on non master branches.

**Comment by nemaniarjun** [comment]
@Asalle here is my interpretation of what the original issue means:
`coala` is not supported on `py2` and `py3.3` so we have a script (.misc/check_unsupported.sh) that checks if running coala on these versions gives the `unsupported message` we expect or not.

Now there are 2 types of `Travis` builds, one which runs on every PR and one which periodically runs on master and updates the badge in the README.

The original issue is to completely remove py2 and py3.3 from PR builds and have these unsupported checks only in the master builds.

This is my interpretation, @jayvdb can you please confirm this? Thanks :)

File: .misc/travis_build_and_test.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+set -e
+
+if [[ ${UNSUPPORTED} == "false" || ${TRAVIS_BRANCH} == "master" ]]; then
+    py.test
```
Comment (jayvdb) [inline .misc/travis_build_and_test.sh]: please ensure your script complies with https://github.com/coala/coala/pull/5243 ; ping @sangamcse

Comment (sangamcse) [reply .misc/travis_build_and_test.sh]: @Asalle, re-push after changing this to 2-space indent. 😃

File: .misc/travis_build_and_test.sh [hunk 11]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+set -e
+
+if [[ ${UNSUPPORTED} == "false" || ${TRAVIS_BRANCH} == "master" ]]; then
```
Comment (jayvdb) [inline .misc/travis_build_and_test.sh]: wow; no way; we can do this via travis rules.
